File: leetcode-04/solutions/python3/0152.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        ans = nums[0]
        max_ = ans
        min_ = ans

        for i in range(1, len(nums)):
            if nums[i] < 0:
                max_, min_ = min_, max_
            max_ = max(nums[i], max_ * nums[i])
            min_ = min(nums[i], min_ * nums[i])
            ans = max(ans, max_)

        return ans

File: leetcode-04/solutions/cpp/0152.cpp
class Solution {
 public:
  int maxProduct(vector<int>& nums) {
    int ans = nums[0];
    int max = ans;
    int min = ans;

    for (int i = 1; i < nums.size(); i++) {
      if (nums[i] < 0) swap(max, min);
      max = std::max(nums[i], max * nums[i]);
      min = std::min(nums[i], min * nums[i]);
      ans = std::max(ans, max);
    }

    return ans;
  }
};